### PR TITLE
Added error handling for all the know errors that MAP STS can return

### PR DESCRIPTION
**Pull request by achievedaniel**
## Summary

Added new error handling for the MAP STS service. There is a few different responses that the service returns depending on what is missing or invalid in the request for more details please check the comments in the related issue

## Related issue(s)
department-of-veterans-affairs/va.gov-team#65130

## Testing done

Spec created that mocks all the different known errors that MAP STS can return

## What areas of the site does it impact?

Impacts the integration with MAP STS

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution

### CODE REVIEW

File: lib/mobile_application_platform/security_token/service.rb [hunk 1]
```diff
@@ -16,9 +16,23 @@ def token(application:, icn:)
...
+                        if e.body.is_a?(Hash) && e.body&.key?(:error)
+                          e.body[:error]
+                        elsif e.body.is_a?(Hash) && e.body&.key?(:error_description)
+                          e.body[:error_description]
+                        else
+                          'unknown'
+                        end
+                      else
+                        'unknown'
+                      end
```
Comment (dickdavis) [inline lib/mobile_application_platform/security_token/service.rb:32]: I think this is a great opportunity for pattern matching. Here's what this would look like, using pattern matching to clean it up:

```ruby
description = case e
              in _msg, 500, _body
                'service returned Internal Server error most likely due to an invalid ICN'
              in _msg, (400|401), { error: }
                error
              in _msg, (400|401), { error_description: }
                error_description
              else
                'unknown'
              end
```

In order for this to work, the error must respond to `deconstruct`. This is pretty easy to do, won't mess anything up, and will allow other devs to pattern match off instances of this error class in the future. You can just implement the `deconstruct` method in the `Common::Client::Errors::ClientError` class like so:

```ruby
module Common
  module Client
    module Errors
      class ClientError < Error
      ...
      def deconstruct
        [message, status, body]
      end
    end
  end
end
```

File: spec/lib/mobile_application_platform/security_token/service_spec.rb [hunk 2]
```diff
@@ -12,23 +12,70 @@
     let(:log_prefix) { '[MobileApplicationPlatform][SecurityToken][Service]' }
 
     shared_examples 'STS token request' do
-      context 'when an issue occurs with the client request' do
-        let(:expected_error) { Common::Client::Errors::ClientError }
-        let(:expected_error_message) do
-          "#{log_prefix} token failed, client error, status: #{status}, description: #{description}," \
-            " application: #{application}, icn: #{icn}"
+      shared_examples 'Client Errors' do
```
Comment (dickdavis) [inline spec/lib/mobile_application_platform/security_token/service_spec.rb:15]: This is a nitpick, so feel free to ignore, but the shared examples need to have a more descriptive name in my opinion. Something that reads like natural language from the spec output, like "a request with client errors".

**Review by dickdavis — CHANGES_REQUESTED** [review]
Specs look good, but I think the variable assignment to `description` could be cleaned up. Happy to pair on this if you like.

File: lib/mobile_application_platform/security_token/service.rb [hunk 3]
```diff
@@ -16,9 +16,19 @@ def token(application:, icn:)
...
+        description = case e
+                      in _msg, 500, _body
+                        'service returned Internal Server error most likely due to an invalid ICN'
+                      in _msg, (400|401), { error: }
+                        error
+                      in _msg, (400|401), { error_description: }
+                        error_description
+                      else
+                        'unknown'
+                      end
```
Comment (rileyanderson) [inline lib/mobile_application_platform/security_token/service.rb:28]: I think would be cleaner and more maintainable to create new error classes in [here](https://github.com/department-of-veterans-affairs/vets-api/blob/master/lib/mobile_application_platform/security_token/errors.rb) for these different scenarios and handle the logic of which error to raise in a base error class. Here is an example of what I'm talking about - [error.rb](https://github.com/department-of-veterans-affairs/sign-in-service-rb/blob/main/lib/sign_in_service/error.rb) (won't be exactly like this)

Comment (dickdavis) [reply lib/mobile_application_platform/security_token/service.rb:28]: Nice, I agree. Plus, you get the added benefit of raising an error more descriptive than the generic client error.

File: modules/claims_api/spec/requests/v1/claims_request_spec.rb [hunk 4]
```diff
@@ -44,6 +44,7 @@
   before do
     stub_poa_verification
     stub_mpi
+    stub_jwt_valid_token_decode
```
Comment (bosawt) [inline modules/claims_api/spec/requests/v1/claims_request_spec.rb:47]: I'm not sure why this was added for all of these files but it's certainly not necessary. These changes are unrelated to these files

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb [hunk 5]
```diff
@@ -5,24 +5,27 @@ module V0
     class FacilitiesInfoController < ApplicationController
       SORT_METHODS = %w[home current alphabetical appointments].freeze
 
+      before_action :validate_sort_method_inclusion!, only: %i[schedulable]
```
Comment (bosawt) [inline modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:8]: Was it possible you pulled in unintentional changes? These don't seem related at all

File: lib/mobile_application_platform/security_token/service.rb [hunk 6]
```diff
@@ -16,9 +16,19 @@ def token(application:, icn:)
         Rails.logger.info("#{config.logging_prefix} token success, application: #{application}, icn: #{icn}")
         parse_response(response, application, icn)
       rescue Common::Client::Errors::ClientError => e
+        description = case e
+                      in _msg, 500, _body
+                        'service returned Internal Server error most likely due to an invalid ICN'
```
Comment (bosawt) [inline lib/mobile_application_platform/security_token/service.rb:21]: I wouldn't interpret errors 'most likely due to...' doesn't cut it as an error description (for example, what happens when we get this error and it's not due to an invalid ICN?). Just extract the error description and return that in an error

File: lib/mobile_application_platform/security_token/service.rb [hunk 7]
```diff
@@ -16,9 +16,19 @@ def token(application:, icn:)
...
+                      in _msg, (400|401), { error: }
+                        error
+                      in _msg, (400|401), { error_description: }
```
Comment (bosawt) [inline lib/mobile_application_platform/security_token/service.rb:24]: the pattern matching here is clever, but I'd argue we're making this unnecessarily brittle, conveying the errors we know (which we probably don't need much information about), and losing information about the errors we don't know (which we probably do need information about). Also if we get a status other than a `500/400/401`, the only guidance we'll get here is an `unknown` description, which probably is less useful than letting the original error itself be raised

Comment (bosawt) [reply lib/mobile_application_platform/security_token/service.rb:24]: I would review the different ways this service appears to fail and just accommodate those, don't worry about `error_description` (especially since I'm pretty sure that was my mistake and will never be returned by this service).

Basically I'd just extract the status and plop it in the new error description, extract the description and plot it in the new error description, add additional context, and call it a day

**Review by bosawt — CHANGES_REQUESTED** [review]
Some changes requested to make this work in less of a brittle way. 

Also there were many changes included in this PR that I don't think were part of this work, I believe a failed merge may have occurred?

File: modules/claims_api/spec/requests/v1/claims_request_spec.rb (hunk 4 above)
Comment (dickdavis) [reply modules/claims_api/spec/requests/v1/claims_request_spec.rb:47]: I think he needs to rebase and drop some commits. It looks like he merged.

File: lib/mobile_application_platform/security_token/service.rb (hunk 6 above)
Comment (achievedaniel) [reply lib/mobile_application_platform/security_token/service.rb:21]: Yeah I understand and I dont like the 'most likely' error description either however most of the errors coming from the API are useless and provide even less info. On this example it just return a 500 with Internal Server Error. With no other explanation. Some other errors are even worse and just return and empty body

File: lib/mobile_application_platform/security_token/service.rb (hunk 7 above)
Comment (achievedaniel) [reply lib/mobile_application_platform/security_token/service.rb:24]: Yes I do understand if the API had a better error handling it would be better. I was just trying to giving some light to the errors the API may have.

File: modules/claims_api/spec/requests/v1/claims_request_spec.rb (hunk 4 above)
Comment (achievedaniel) [reply modules/claims_api/spec/requests/v1/claims_request_spec.rb:47]: Yeah I don't know what happened looks like I added commits from someone else on accident. I will just create a new branch and a new pull request.

File: modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb (hunk 5 above)
Comment (dickdavis) [reply modules/mobile/app/controllers/mobile/v0/facilities_info_controller.rb:8]: You should just be able to rebase and drop the other commits.

```
git rebase origin/master -i
```

Then, change `pick` to `drop` for all of the commits that aren't yours.

File: lib/common/client/errors.rb [hunk 8]
```diff
@@ -14,6 +14,10 @@ def initialize(message = nil, status = nil, body = nil)
           @status = status
           @body = body
         end
+
+        def deconstruct
+          [message, status, body]
+        end
```
Comment (dickdavis) [inline lib/common/client/errors.rb:20]: ✂️ You can remove this line since we decided against pattern matching the response.

**Review by dickdavis — CHANGES_REQUESTED** [review]

File: lib/mobile_application_platform/security_token/service.rb [hunk 9]
```diff
@@ -16,14 +16,29 @@ def token(application:, icn:)
...
+        context = {
+          error_description: parse_body['error_description'],
+          error: parse_body['error'],
+          error_uri: parse_body['error_uri']
+        }.compact
+        raise e, "#{config.logging_prefix} Token failed, client error, status: #{status}," \
+                 " application: #{application}, icn: #{icn}, context: #{context}"
+      end
+
+      def parse_json(to_parse)
```
Comment (bosawt) [inline lib/mobile_application_platform/security_token/service.rb:36]: I probably wouldn't `rescue` an error from `JSON.parse` here, theoretically we should always expect parseable JSON, and if we don't get it, we want to raise an error and get it to Sentry so that we can track it. This is an example of eating the error and removing actionable context

File: spec/lib/mobile_application_platform/security_token/service_spec.rb [hunk 10]
```diff
@@ -12,23 +12,73 @@
...
+        context 'when an issue occurs with the client request' do
+          let(:expected_error) { Common::Client::Errors::ClientError }
+          let(:expected_error_message) do
+            "#{log_prefix} Token failed, client error, status: #{status}," \
+              " application: #{application}, icn: #{icn}, context: #{expected_error_body}"
+          end
+
+          before do
+            allow_any_instance_of(described_class).to receive(:perform).and_raise(raised_error)
```
Comment (bosawt) [inline spec/lib/mobile_application_platform/security_token/service_spec.rb:24]: I know the previous specs just raised errors to simulate failed responses, but now that we're doing this 'properly' we should have these responses in VCR and replay them in these specs. Refer to `spec/lib/mobile_application_platform/sign_up/service_spec.rb` for an example of what this looks like

File: lib/common/client/errors.rb (hunk 8 above)
Comment (bosawt) [reply lib/common/client/errors.rb:20]: This is accurate, no need for this

**Comment by bosawt** [comment]
This work is completed here: https://github.com/department-of-veterans-affairs/vets-api/pull/14246
